Fix AFTER_SETUP event not being able to render

JustRed23 opened the pull request:
Due to 1.21.2 changing the way how level rendering works, adding a `RenderSystem.clear` after the `AFTER_SETUP` event, rendering is not possible with this event, so I moved the event to a later point after the clear call but before the terrain rendering

JustRed23 committed:
Add BEFORE_TERRAIN world render event

JustRed23 committed:
Add 1.21.2 rendering issue warning

Alexander01998 commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/mixin/client/rendering/WorldRendererMixin.java:
```diff
@@ -83,6 +83,19 @@ private void afterTerrainSetup(Camera camera, Frustum frustum, boolean hasForced
 		WorldRenderEvents.AFTER_SETUP.invoker().afterSetup(context);
 	}
 
+	@Inject(
+			method = "method_62214",
+			at = @At(
+				value = "INVOKE",
+				target = "Lnet/minecraft/util/profiler/Profiler;push(Ljava/lang/String;)V",
+				ordinal = 0,
+				shift = Shift.AFTER
+			)
+	)
```
This points to `profiler.push("terrain");` if I'm not mistaken? Might be a good idea to note that with a comment since it's not obvious from the code.

JustRed23 replied:
It does, I'll add a comment

JustRed23 committed:
Add comment to injection point

watertrainer commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/WorldRenderEvents.java:
```diff
@@ -68,13 +69,26 @@ private WorldRenderEvents() { }
 	 *
 	 * <p>Use for setup of state that depends on view frustum.
+	 *
+	 * <b>Can NOT be used to render anything after MC 1.21.2 due to a glClear call, use {@link WorldRenderEvents#BEFORE_TERRAIN} instead.
```
building fails for me because of a missing closing tag for the bold text here, works after adding it

JustRed23 replied:
Weird.. the rest of the docs also don't use closing tags

Juuxel replied:
The docs don't use closing tags for `<p>`. That format is pretty conventional for javadocs, and I think it's a special case in the Checkstyle rules. (In valid HTML, the end tag `</b>` can't be omitted but `</p>` can in this case.)

AsynchronousAI replied:
I got this issue also, adding `</b>` fixed this.

JustRed23 committed:
close b tag

AsynchronousAI commented:
LGTM!

modmuss50 reviewed (changes requested):
It was pointed out to me on discord that the old API is broken, this should be fixed to work as it did it before (or as close to it). There is no point adding a new api and leaving a broken API in its place.

JustRed23 committed:
Undo all new event changes

JustRed23 committed:
Move AFTER_SETUP event to new mixin

JustRed23 committed:
Remove event interface

JustRed23 commented:
> It was pointed out to me on discord that the old API is broken, this should be fixed to work as it did it before (or as close to it). There is no point adding a new api and leaving a broken API in its place.

Removed new event, fixed existing event

Chocohead commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/mixin/client/rendering/WorldRendererMixin.java:
```diff
@@ -80,6 +80,18 @@ private void beforeRender(ObjectAllocator objectAllocator, RenderTickCounter tic
+	@Inject(
+			method = "method_62214",
+			at = @At(
+				value = "INVOKE",
```
Would this not be a good place to use `INVOKE_STRING`? That way this can specifically target the `terrain` profiler push call, even if it gets moved around

JustRed23 committed:
Refine mixin
